File: c#/c_recursive_calc/c_recursive_calc/Lexer.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace c_recursive_calc
{
    class Lexer : ILexer
    {
        public static string[] operators = new []{ "+", "-", "*", "/", "^", "(", ")"};
        public List<Token> Lex(string str)
        {
            List<Token> TokenList = new List<Token>();
            int iter = 0;
            while(true)
            {
                while (iter < str.Length && Char.IsWhiteSpace(str[iter]))
                    iter++;
                Token token = new Token();
                if (iter >= str.Length)
                {
                    TokenList.Add(token);
                    break;
                }
                else if (Char.IsNumber(str[iter]))
                {
                    while (iter < str.Length && Char.IsNumber(str[iter]))
                        token.Text += str[iter++];
                    token.Id = Token.TokenId.Number;
                }
                else
                {
                    token.Text += str[iter++];
                    if (operators.Contains(token.Text))
                        token.Id = Token.TokenId.Operator;
                    else
                        token.Id = Token.TokenId.Unknown;
                }
               TokenList.Add(token);
            }
            return TokenList;
        }
    }
}

File: php/calc_parser/Lexer.php
<?php
include "ILexer.php";
include "Token.php";
/**
 * Created by PhpStorm.
 * User: SRigh
 * Date: 11.09.2015
 * Time: 13:02
 */

class Lexer implements ILexer
{
    private $operators = ["$","+", "-", "*", "/", "^", "(", ")"];
    public static $TokenList = array();

    public function Lex($str)
    {
        $iter = 0;
        while(true)
        {
            while($iter < strlen($str) && $str[$iter] == " ")
                $iter++;
            $token = new Token();
            if($iter >= strlen($str))
            {
                $TokenList[] = $token;
                break;
            }
            elseif(is_numeric($str[$iter]))
            {
                while($iter < strlen($str) && is_numeric($str[$iter]))
                {
                    $token->text .= $str[$iter++];
                    $token->id = $token->tokenId['number'];
                }
            }
            else
            {
                $token->text = $str[$iter++];
                if(array_search($token->text,$this->operators))
                    $token->id = $token->tokenId['operator'];
                else
                    $token->id = $token->tokenId['unknown'];
            }
            self::$TokenList[] = $token;
        }
//        print_r($TokenList);
        return self::$TokenList;
    }
}
